jdbc-yugabytedb/src/main/java/org/postgresql/Driver.java

```diff
@@ -450,12 +451,89 @@ public Connection getResult(long timeout) throws SQLException {
    * thread without enforcing a timeout, regardless of any timeout specified in the properties.
    *
    * @param url the original URL
-   * @param props the parsed/defaulted connection properties
+   * @param properties the parsed/defaulted connection properties
    * @return a new connection
    * @throws SQLException if the connection could not be made
    */
-  private static Connection makeConnection(String url, Properties props) throws SQLException {
-    return new PgConnection(hostSpecs(props), user(props), database(props), props, url);
+  private static Connection makeConnection(String url, Properties properties) throws SQLException {
+    LoadBalanceProperties lbprops = new LoadBalanceProperties(url, properties);
+    if (lbprops.hasLoadBalance()) {
+      Connection conn = getConnectionBalanced(lbprops);
+      if (conn != null) {
+        return conn;
+      } else {
+        LOGGER.log(Level.WARNING, "Failed to apply load balance. Trying normal connection");
+        // but remember to use stripped url and props
+        url = lbprops.getStrippedURL();
+        properties = lbprops.getStrippedProperties();
+      }
+    }
+    return new PgConnection(hostSpecs(properties), user(properties), database(properties), properties, url);
+  }
+
+  private static Connection getConnectionBalanced(LoadBalanceProperties lbprops) throws SQLException {
+    ClusterAwareLoadBalancer loadBalancer = lbprops.getAppropriateLoadBalancer();
+    Properties props = lbprops.getStrippedProperties();
+    String url = lbprops.getStrippedURL();
+    Set<String> unreachableHosts = loadBalancer.getUnreachableHosts();
+    List<String> failedHosts = new ArrayList<>(unreachableHosts);
+    String chosenHost = loadBalancer.getLeastLoadedServer(failedHosts);
+    Connection newConnection = null;
+    Connection controlConnection = null;
+    SQLException firstException = null;
+    if (chosenHost == null) {
+      HostSpec[] hspec = hostSpecs(lbprops.getOriginalProperties());
+      controlConnection = new PgConnection(
+        hspec, user(lbprops.getOriginalProperties()),
+          database(lbprops.getOriginalProperties()), props, url);
+      try {
+        LOGGER.log(Level.FINE, "refreshing server list from {0}", hspec[0].getHost());
+        loadBalancer.refresh(controlConnection);
+        controlConnection.close();
+      } catch (PSQLException ex) {
+        if (PSQLState.UNDEFINED_FUNCTION.getState().equals(ex.getSQLState())) {
+          return null;
+        }
+        throw ex;
+      }
+    }
+    chosenHost = loadBalancer.getLeastLoadedServer(failedHosts);
```

Comment:
This seems to be the same call as that on line 480.
why is this needed ?

Reply:
I think this line should be executed when chosenHost was null.
The expectation is that, after refreshing load balancer, we can pick the least loaded server.
Maybe move this call inside the if block after line 498.

Reply:
Yes. You are right. Thanks.